### ISSUE TITLE
`SequenceGenerator` is somewhat flaky on all platforms in `TestTranslation::test_multilingual_translation_latent_depth`

### ISSUE THREAD

**Issue by erip** [pytorch/fairseq#2749]
## 🐛 Bug

Sometimes [this](https://github.com/pytorch/fairseq/blob/master/fairseq/sequence_generator.py#L416) assertion is triggered in the test TestTranslation::test_multilingual_translation_latent_depth. The reason seems to be related to the fact that [this condition](https://github.com/pytorch/fairseq/blob/master/fairseq/sequence_generator.py#L391) is sometimes not met, meaning that the num remaining sentences is never decremented. 

This is likely due to some beam search non-determinism.

### To Reproduce

Steps to reproduce the behavior (**always include the command you ran**):

1. Clone and install
2. Install pytest
3. run `pytest tests/test_binaries.py::TestTranslation::test_multilingual_translation_latent_depth` several times until observe failure

### Expected behavior

The test should not be flaky. 

### Environment

 - fairseq Version (e.g., 1.0 or master): master
 - PyTorch Version (e.g., 1.0) 1.6.0
 - OS (e.g., Linux): OS X, Windows
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source): normal install commands
 - Python version: 3.6
 - CUDA/cuDNN version: n/a
 - GPU models and configuration: n/a
 - Any other relevant information: n/a

**Comment by erip** [comment]
Something interesting is when it fails, lprobs from [here](https://github.com/pytorch/fairseq/blob/master/fairseq/sequence_generator.py#L304) becomes a tensor completely full of nans.

**Comment by erip** [comment]
On second thought, it is likely due to the translation data generation process. Could it be the case that the data generated by `create_dummy_data` is numerically unstable? Or that some process in the pipeline is? It's odd that this is the only flaky test since everything uses sequence generators...

**Comment by myleott** [comment]
Yeah there's something weird about it. I think it's because it's doing some gumbel softmax stuff. It seemed it's only flakey on MacOS, so I was going to skip the test if `sys.platform.lower() == "darwin"`. Have you seen flakiness on linux? Perhaps we just disable for darwin + windows?

**Comment by erip** [comment]
I haven't tested on Linux, but it seems like CI seems clear on Ubuntu. I am happy to put a skip around it for now with a TODO.

**Comment by myleott** [comment]
I disabled it on macOS, seems okay now? I’ll follow up with the author of that test about it.
